File: cookbook/javascript/template_building.js
/**
 * Template Building Cookbook - TypeScript/JavaScript
 * 
 * Examples of building custom templates with the Hopx SDK
 */

const { Template, waitForPort, waitForURL } = require('@hopx-ai/sdk');

// =============================================================================
// Example 1: Simple Python Web App
// =============================================================================

async function example1_pythonWebApp() {
  console.log('Example 1: Python Flask Web App\n');
  
  const template = new Template()
    .fromPythonImage('3.11')
    .copy('app/', '/app/')
    .pipInstall()  // Installs from requirements.txt
    .setEnv('FLASK_APP', 'app.py')
    .setEnv('FLASK_ENV', 'production')
    .setStartCmd('python /app/app.py', waitForPort(5000));
  
  const result = await Template.build(template, {
    alias: 'my-flask-app',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 2,
    memory: 2048,
    diskGB: 10,
    onLog: (log) => console.log(`[${log.level}] ${log.message}`),
    onProgress: (progress) => console.log(`Progress: ${progress}%`)
  });
  
  console.log(`✅ Template built: ${result.templateID}`);
  
  // Create VM from template
  const vm = await result.createVM({
    alias: 'flask-instance-1'
  });
  
  console.log(`✅ VM created: ${vm.ip}`);
  
  // Cleanup
  await vm.delete();
}

// =============================================================================
// Example 2: Node.js API Server
// =============================================================================

async function example2_nodejsAPI() {
  console.log('Example 2: Node.js Express API\n');
  
  const template = new Template()
    .fromNodeImage('18-alpine')
    .copy('package.json', '/app/package.json')
    .copy('package-lock.json', '/app/package-lock.json')
    .copy('src/', '/app/src/')
    .setWorkdir('/app')
    .npmInstall()  // Installs from package.json
    .setEnv('NODE_ENV', 'production')
    .setEnv('PORT', '3000')
    .setStartCmd('node src/server.js', waitForPort(3000));
  
  const result = await Template.build(template, {
    alias: 'express-api',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 4,
    memory: 4096,
    diskGB: 20
  });
  
  console.log(`✅ Template: ${result.templateID}`);
  
  // Create multiple instances
  const instances = await Promise.all([
    result.createVM({ alias: 'api-1' }),
    result.createVM({ alias: 'api-2' }),
    result.createVM({ alias: 'api-3' })
  ]);
  
  console.log(`✅ Created ${instances.length} instances`);
}

// =============================================================================
// Example 3: Full Stack App (Python + Redis)
// =============================================================================

async function example3_fullStack() {
  console.log('Example 3: Full Stack with Redis\n');
  
  const template = new Template()
    .fromUbuntuImage('22.04')
    .aptInstall(['python3-pip', 'redis-server'])
    .copy('backend/', '/app/')
    .pipInstall(['flask', 'redis', 'celery'])
    .setEnv('REDIS_URL', 'redis://localhost:6379')
    .setEnv('FLASK_APP', 'app.py')
    .setWorkdir('/app')
    .setStartCmd(
      'redis-server --daemonize yes && python app.py',
      waitForPort(8000)
    );
  
  const result = await Template.build(template, {
    alias: 'fullstack-app',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 4,
    memory: 4096,
    diskGB: 20
  });
  
  console.log(`✅ Full stack template ready`);
}

// =============================================================================
// Example 4: Database Server (PostgreSQL)
// =============================================================================

async function example4_database() {
  console.log('Example 4: PostgreSQL Database\n');
  
  const template = new Template()
    .fromUbuntuImage('22.04')
    .aptInstall(['postgresql', 'postgresql-contrib'])
    .setEnv('POSTGRES_PASSWORD', 'secret')
    .setEnv('POSTGRES_DB', 'myapp')
    .setEnv('POSTGRES_USER', 'admin')
    .setStartCmd(
      'service postgresql start',
      waitForPort(5432)
    );
  
  const result = await Template.build(template, {
    alias: 'postgres-server',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 2,
    memory: 4096,
    diskGB: 50
  });
  
  console.log(`✅ Database template: ${result.templateID}`);
}

// =============================================================================
// Example 5: ML/AI Workload
// =============================================================================

async function example5_mlWorkload() {
  console.log('Example 5: ML/AI with TensorFlow\n');
  
  const template = new Template()
    .fromPythonImage('3.11')
    .aptInstall(['git', 'curl', 'wget'])
    .pipInstall(['tensorflow', 'numpy', 'pandas', 'jupyter'])
    .copy('notebooks/', '/notebooks/')
    .copy('models/', '/models/')
    .setEnv('JUPYTER_TOKEN', 'secret123')
    .setWorkdir('/notebooks')
    .setStartCmd(
      'jupyter notebook --ip=0.0.0.0 --port=8888 --no-browser --allow-root',
      waitForPort(8888)
    );
  
  const result = await Template.build(template, {
    alias: 'ml-workload',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 8,
    memory: 16384,
    diskGB: 100
  });
  
  console.log(`✅ ML template ready`);
}

// =============================================================================
// Example 6: Development Environment
// =============================================================================

async function example6_devEnvironment() {
  console.log('Example 6: Development Environment\n');
  
  const template = new Template()
    .fromUbuntuImage('22.04')
    .aptInstall(['git', 'curl', 'wget', 'vim', 'build-essential'])
    .runCmd('curl -fsSL https://deb.nodesource.com/setup_18.x | bash -')
    .aptInstall(['nodejs', 'python3-pip'])
    .pipInstall(['ipython', 'pytest', 'black'])
    .npmInstall(['typescript', 'eslint', 'prettier'])
    .setEnv('EDITOR', 'vim')
    .setWorkdir('/workspace');
  
  const result = await Template.build(template, {
    alias: 'dev-environment',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 4,
    memory: 8192,
    diskGB: 50
  });
  
  console.log(`✅ Dev environment template ready`);
}

// =============================================================================
// Example 7: Git Clone + Build
// =============================================================================

async function example7_gitClone() {
  console.log('Example 7: Clone GitHub Repo & Build\n');
  
  const template = new Template()
    .fromNodeImage('18')
    .aptInstall(['git'])
    .gitClone('https://github.com/example/repo.git', '/app')
    .setWorkdir('/app')
    .npmInstall()
    .runCmd('npm run build')
    .setEnv('NODE_ENV', 'production')
    .setStartCmd('npm start', waitForPort(3000));
  
  const result = await Template.build(template, {
    alias: 'github-app',
    apiKey: process.env.HOPX_API_KEY
  });
  
  console.log(`✅ Template from GitHub: ${result.templateID}`);
}

// =============================================================================
// Example 8: Multi-Language Environment
// =============================================================================

async function example8_multiLanguage() {
  console.log('Example 8: Multi-Language (Python + Node + Go)\n');
  
  const template = new Template()
    .fromUbuntuImage('22.04')
    // Python
    .aptInstall(['python3-pip'])
    .pipInstall(['flask', 'requests'])
    // Node.js
    .runCmd('curl -fsSL https://deb.nodesource.com/setup_18.x | bash -')
    .aptInstall(['nodejs'])
    // Go
    .runCmd('wget https://go.dev/dl/go1.21.0.linux-amd64.tar.gz')
    .runCmd('tar -C /usr/local -xzf go1.21.0.linux-amd64.tar.gz')
    .setEnv('PATH', '/usr/local/go/bin:$PATH')
    .setWorkdir('/workspace');
  
  const result = await Template.build(template, {
    alias: 'multi-lang-env',
    apiKey: process.env.HOPX_API_KEY,
    cpu: 4,
    memory: 8192,
    diskGB: 30
  });
  
  console.log(`✅ Multi-language environment ready`);
}

// =============================================================================
// Example 9: Caching Strategy
// =============================================================================

async function example9_cachingStrategy() {
  console.log('Example 9: Efficient Caching\n');
  
  // First build - installs dependencies
  const template1 = new Template()
    .fromPythonImage('3.11')
    .pipInstall(['flask', 'gunicorn', 'redis', 'celery'])
    .copy('app/', '/app/')
    .skipCache()  // Force rebuild
    .setStartCmd('gunicorn app:app', waitForPort(8000));
  
  const result1 = await Template.build(template1, {
    alias: 'app-v1',
    apiKey: process.env.HOPX_API_KEY
  });
  
  console.log(`✅ V1 built: ${result1.templateID}`);
  
  // Second build - will use cache for dependencies
  const template2 = new Template()
    .fromPythonImage('3.11')
    .pipInstall(['flask', 'gunicorn', 'redis', 'celery'])  // Cached!
    .copy('app/', '/app/')  // New files
    .setStartCmd('gunicorn app:app', waitForPort(8000));
  
  const result2 = await Template.build(template2, {
    alias: 'app-v2',
    apiKey: process.env.HOPX_API_KEY
  });
  
  console.log(`✅ V2 built (with cache): ${result2.templateID}`);
}

// =============================================================================
// Example 10: Health Check Patterns
// =============================================================================

async function example10_healthChecks() {
  console.log('Example 10: Different Health Check Types\n');
  
  // Port-based health check
  const template1 = new Template()
    .fromPythonImage('3.11')
    .pipInstall(['flask'])
    .setStartCmd('flask run', waitForPort(5000));
  
  // URL-based health check
  const template2 = new Template()
    .fromNodeImage('18')
    .npmInstall(['express'])
    .setStartCmd('node server.js', waitForURL('http://localhost:3000/health'));
  
  // File-based health check
  const template3 = new Template()
    .fromUbuntuImage('22.04')
    .runCmd('touch /tmp/ready')
    .setStartCmd('/bin/bash', waitForFile('/tmp/ready'));
  
  // Process-based health check
  const template4 = new Template()
    .fromUbuntuImage('22.04')
    .aptInstall(['nginx'])
    .setStartCmd('nginx -g "daemon off;"', waitForProcess('nginx'));
  
  // Command-based health check
  const template5 = new Template()
    .fromUbuntuImage('22.04')
    .aptInstall(['postgresql'])
    .setStartCmd('service postgresql start', waitForCommand('pg_isready'));
  
  console.log('✅ 5 different health check patterns defined');
}

// =============================================================================
// Run Examples
// =============================================================================

async function runAll() {
  try {
    await example1_pythonWebApp();
    await example2_nodejsAPI();
    await example3_fullStack();
    await example4_database();
    await example5_mlWorkload();
    await example6_devEnvironment();
    await example7_gitClone();
    await example8_multiLanguage();
    await example9_cachingStrategy();
    await example10_healthChecks();
    
    console.log('\n✨ All examples completed!');
  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

// Run if called directly
if (require.main === module) {
  runAll();
}

module.exports = {
  example1_pythonWebApp,
  example2_nodejsAPI,
  example3_fullStack,
  example4_database,
  example5_mlWorkload,
  example6_devEnvironment,
  example7_gitClone,
  example8_multiLanguage,
  example9_cachingStrategy,
  example10_healthChecks
};

File: cookbook/python/template_building.py
#!/usr/bin/env python3
"""
Template Building Cookbook - Python

Examples of building custom templates with the Hopx SDK
"""

import os
import asyncio
from hopx_ai import Template, wait_for_port, wait_for_url, wait_for_process
from hopx_ai.template import BuildOptions, CreateVMOptions


# =============================================================================
# Example 1: Simple Python Web App
# =============================================================================

async def example1_python_web_app():
    """Build a Flask web application template"""
    print("Example 1: Python Flask Web App\n")
    
    template = (
        Template()
        .from_python_image("3.11")
        .copy("app/", "/app/")
        .pip_install()  # Installs from requirements.txt
        .set_env("FLASK_APP", "app.py")
        .set_env("FLASK_ENV", "production")
        .set_start_cmd("python /app/app.py", wait_for_port(5000))
    )
    
    result = await Template.build(
        template,
        BuildOptions(
            alias="my-flask-app",
            api_key=os.environ["HOPX_API_KEY"],
            cpu=2,
            memory=2048,
            disk_gb=10,
            on_log=lambda log: print(f"[{log['level']}] {log['message']}"),
            on_progress=lambda p: print(f"Progress: {p}%")
        )
    )
    
    print(f"✅ Template built: {result.template_id}")
    
    # Create VM from template
    vm = await result.create_vm(CreateVMOptions(alias="flask-instance-1"))
    print(f"✅ VM created: {vm.ip}")
    
    # Cleanup
    await vm.delete()


# =============================================================================
# Example 2: Data Science Environment
# =============================================================================

async def example2_data_science():
    """Build a Jupyter notebook environment"""
    print("Example 2: Data Science with Jupyter\n")
    
    template = (
        Template()
        .from_python_image("3.11")
        .pip_install([
            "jupyter",
            "numpy",
            "pandas",
            "matplotlib",
            "scikit-learn",
            "tensorflow"
        ])
        .copy("notebooks/", "/notebooks/")
        .set_env("JUPYTER_TOKEN", "mysecret")
        .set_workdir("/notebooks")
        .set_start_cmd(
            "jupyter notebook --ip=0.0.0.0 --port=8888 --no-browser --allow-root",
            wait_for_port(8888)
        )
    )
    
    result = await Template.build(
        template,
        BuildOptions(
            alias="jupyter-datascience",
            api_key=os.environ["HOPX_API_KEY"],
            cpu=8,
            memory=16384,
            disk_gb=100
        )
    )
    
    print(f"✅ Data science environment: {result.template_id}")


# =============================================================================
# Example 3: Microservices Stack
# =============================================================================

async def example3_microservices():
    """Build a microservices environment"""
    print("Example 3: Microservices Stack\n")
    
    template = (
        Template()
        .from_ubuntu_image("22.04")
        .apt_install(["python3-pip", "redis-server", "nginx"])
        .copy("services/", "/services/")
        .pip_install(["flask", "fastapi", "uvicorn", "redis"])
        .set_env("SERVICE_PORT", "8000")
        .set_env("REDIS_PORT", "6379")
        .set_workdir("/services")
        .set_start_cmd(
            "redis-server --daemonize yes && uvicorn main:app --host 0.0.0.0",
            wait_for_port(8000)
        )
    )
    
    result = await Template.build(
        template,
        BuildOptions(
            alias="microservices-stack",
            api_key=os.environ["HOPX_API_KEY"],
            cpu=4,
            memory=4096,
            disk_gb=30
        )
    )
    
    print(f"✅ Microservices template: {result.template_id}")


# =============================================================================
# Example 4: Development Environment
# =============================================================================

async def example4_dev_environment():
    """Build a complete development environment"""
    print("Example 4: Development Environment\n")
    
    template = (
        Template()
        .from_ubuntu_image("22.04")
        .apt_install([
            "git", "curl", "wget", "vim", "build-essential",
            "python3-pip", "python3-dev"
        ])
        .run_cmd("curl -fsSL https://deb.nodesource.com/setup_18.x | bash -")
        .apt_install(["nodejs"])
        .pip_install(["ipython", "pytest", "black", "mypy"])
        .npm_install(["typescript", "eslint", "prettier"])
        .set_env("EDITOR", "vim")
        .set_workdir("/workspace")
    )
    
    result = await Template.build(
        template,
        BuildOptions(
            alias="dev-environment",
            api_key=os.environ["HOPX_API_KEY"],
            cpu=4,
            memory=8192,
            disk_gb=50
        )
    )
    
    print(f"✅ Dev environment: {result.template_id}")


# =============================================================================
# Example 5: Clone & Build from GitHub
# =============================================================================

async def example5_github_clone():
    """Clone a GitHub repo and build it"""
    print("Example 5: Clone from GitHub\n")
    
    template = (
        Template()
        .from_node_image("18")
        .apt_install(["git"])
        .git_clone("https://github.com/example/repo.git", "/app")
        .set_workdir("/app")
        .npm_install()
        .run_cmd("npm run build")
        .set_env("NODE_ENV", "production")
        .set_start_cmd("npm start", wait_for_port(3000))
    )
    
    result = await Template.build(
        template,
        BuildOptions(
            alias="github-app",
            api_key=os.environ["HOPX_API_KEY"]
        )
    )
    
    print(f"✅ GitHub app template: {result.template_id}")


# =============================================================================
# Example 6: Different Ready Checks
# =============================================================================

async def example6_ready_checks():
    """Demonstrate all ready check types"""
    print("Example 6: Different Ready Check Types\n")
    
    # Port check
    template_port = (
        Template()
        .from_python_image("3.11")
        .pip_install(["flask"])
        .set_start_cmd("flask run", wait_for_port(5000, timeout_seconds=60000))
    )
    
    # URL check
    template_url = (
        Template()
        .from_node_image("18")
        .npm_install(["express"])
        .set_start_cmd(
            "node server.js",
            wait_for_url("http://localhost:3000/health", timeout_seconds=60000)
        )
    )
    
    # Process check
    template_process = (
        Template()
        .from_ubuntu_image("22.04")
        .apt_install(["nginx"])
        .set_start_cmd(
            "nginx -g 'daemon off;'",
            wait_for_process("nginx", timeout_seconds=30000)
        )
    )
    
    print("✅ Defined 3 templates with different ready checks")


# =============================================================================
# Example 7: Resource Optimization
# =============================================================================

async def example7_resource_optimization():
    """Build templates with different resource profiles"""
    print("Example 7: Resource Optimization\n")
    
    # Light template (minimal resources)
    template_light = (
        Template()
        .from_python_image("3.11-slim")
        .pip_install(["flask"])
        .copy("app/", "/app/")
        .set_start_cmd("python /app/app.py", wait_for_port(5000))
    )
    
    result_light = await Template.build(
        template_light,
        BuildOptions(
            alias="light-app",
            api_key=os.environ["HOPX_API_KEY"],
            cpu=1,
            memory=512,
            disk_gb=5
        )
    )
    
    # Heavy template (ML workload)
    template_heavy = (
        Template()
        .from_python_image("3.11")
        .pip_install(["tensorflow", "torch", "transformers"])
        .copy("models/", "/models/")
        .set_start_cmd("python train.py", wait_for_port(8000))
    )
    
    result_heavy = await Template.build(
        template_heavy,
        BuildOptions(
            alias="ml-training",
            api_key=os.environ["HOPX_API_KEY"],
            cpu=16,
            memory=32768,
            disk_gb=200
        )
    )
    
    print(f"✅ Light template: {result_light.template_id}")
    print(f"✅ Heavy template: {result_heavy.template_id}")


# =============================================================================
# Run All Examples
# =============================================================================

async def run_all_examples():
    """Run all cookbook examples"""
    try:
        await example1_python_web_app()
        await example2_data_science()
        await example3_microservices()
        await example4_dev_environment()
        await example5_github_clone()
        await example6_ready_checks()
        await example7_resource_optimization()
        
        print("\n✨ All examples completed!")
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(run_all_examples())

